contracts/NonfungiblePositionManager.sol

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+import {ERC721} from "openzeppelin-contracts/contracts/token/ERC721/ERC721.sol";
+import {INonfungiblePositionManager} from "./interfaces/INonfungiblePositionManager.sol";
+import {BaseLiquidityManagement} from "./base/BaseLiquidityManagement.sol";
+
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {PoolIdLibrary} from "@uniswap/v4-core/src/types/PoolId.sol";
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/src/types/Currency.sol";
+import {CurrencySettleTake} from "./libraries/CurrencySettleTake.sol";
+import {LiquidityRange, LiquidityRangeId, LiquidityRangeIdLibrary} from "./types/LiquidityRange.sol";
+import {BalanceDelta, toBalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+
+import {LiquidityAmounts} from "./libraries/LiquidityAmounts.sol";
+import {TickMath} from "@uniswap/v4-core/src/libraries/TickMath.sol";
+import {StateLibrary} from "@uniswap/v4-core/src/libraries/StateLibrary.sol";
+
+contract NonfungiblePositionManager is INonfungiblePositionManager, BaseLiquidityManagement, ERC721 {
+    using CurrencyLibrary for Currency;
+    using CurrencySettleTake for Currency;
+    using PoolIdLibrary for PoolKey;
+    using LiquidityRangeIdLibrary for LiquidityRange;
+    using StateLibrary for IPoolManager;
+
+    /// @dev The ID of the next token that will be minted. Skips 0
+    uint256 private _nextId = 1;
+
+    struct TokenPosition {
+        address owner;
+        LiquidityRange range;
+    }
+
+    mapping(uint256 tokenId => TokenPosition position) public tokenPositions;
```

Comment:
maybe this will become more clear to me.. but its not clear why we have this mapping and the positions mapping? seems like this just stores LiquidityRange? ... is that really necessary? can we not just pass that in?

and then we only need one mapping:
positions[msg.sender][rangeId]

Reply:
a Position is keyed by its owner and the range

some interfaces (increaseLiquidity, decreaseLiquidity, and collect) operate on tokenId, so we need a way to recover Position from tokenId